crates/net/eth-wire/src/types/broadcast.rs

```diff
@@ -494,140 +489,210 @@ impl HandleMempoolData for NewPooledTransactionHashes68 {
+        PartiallyValidData::from_raw_data_eth66(deduped_data)
     }
 }
 
-impl HandleVersionedMempoolData for NewPooledTransactionHashes68 {
-    fn msg_version(&self) -> EthVersion {
-        EthVersion::Eth68
-    }
+/// Interface for handling mempool message data. Used in various filters in pipelines in
+/// `TransactionsManager` and in queries to `TransactionPool`.
+pub trait HandleMempoolData {
+    /// The announcement contains no entries.
+    fn is_empty(&self) -> bool;
+
+    /// Returns the number of entries.
+    fn len(&self) -> usize;
+
+    /// Retain only entries for which the hash in the entry satisfies a given predicate, return
+    /// the rest.
+    fn retain_by_hash(&mut self, f: impl FnMut(&TxHash) -> bool) -> Self;
 }
 
-impl HandleMempoolData for NewPooledTransactionHashes66 {
+/// Extension of [`HandleMempoolData`] interface, for mempool messages that are versioned.
+pub trait HandleVersionedMempoolData {
+    /// Returns the announcement version, either [`Eth66`](EthVersion::Eth66) or
+    /// [`Eth68`](EthVersion::Eth68).
+    fn msg_version(&self) -> EthVersion;
+}
+
+impl HandleMempoolData for Vec<PooledTransactionsElement> {
     fn is_empty(&self) -> bool {
-        self.0.is_empty()
+        self.is_empty()
     }
 
     fn len(&self) -> usize {
-        self.0.len()
+        self.len()
     }
 
     fn retain_by_hash(&mut self, mut f: impl FnMut(&TxHash) -> bool) -> Self {
         let mut indices_to_remove = vec![];
-        for (i, hash) in self.0.iter().enumerate() {
-            if !f(hash) {
+        for (i, tx) in self.iter().enumerate() {
+            if !f(tx.hash()) {
                 indices_to_remove.push(i);
             }
         }
 
-        let mut removed_hashes = Vec::with_capacity(indices_to_remove.len());
+        let mut removed_txns = Vec::with_capacity(indices_to_remove.len());
 
         for index in indices_to_remove.into_iter().rev() {
-            let hash = self.0.remove(index);
-            removed_hashes.push(hash);
+            let hash = self.remove(index);
+            removed_txns.push(hash);
         }
 
-        Self(removed_hashes)
+        removed_txns
     }
 }
 
-impl HandleVersionedMempoolData for NewPooledTransactionHashes66 {
-    fn msg_version(&self) -> EthVersion {
-        EthVersion::Eth66
-    }
+macro_rules! handle_mempool_data_map_impl {
+    ($data_ty:ty, $(<$generic:ident>)?) => {
+        impl$(<$generic>)? HandleMempoolData for $data_ty {
+            fn is_empty(&self) -> bool {
+                self.data.is_empty()
+            }
+
+            fn len(&self) -> usize {
+                self.data.len()
+            }
+
+            fn retain_by_hash(&mut self, mut f: impl FnMut(&TxHash) -> bool) -> Self {
+                let data = std::mem::take(&mut self.data);
+
+                let (keep, rest) = data.into_iter().partition(|(hash, _)| f(hash));
+
+                self.data = keep;
+
+                Self { data: rest, version: self.version }
```

Comment:
this is relatively expensive because this always allocs while inside the context of the pool

I don't think we need this to return the removed values because these hashes aren't really useful:

https://github.com/paradigmxyz/reth/blob/a6018717da1e112edcaae838c6587d38270bebd0/crates/net/network/src/transactions/mod.rs#L940-L944

we can get the same info here by comparing the length before and after